README.md
dcWordTree
==========

display you tags in an amazing cloud. It uses javascript to generate an Scalable Vector Graphic.


CHANGELOG.md
dcWordTree 0.2.0 - 2015-04-28
==============================
* Add translation
* Fix small issue in admin (missing template).
* Fix deactivation of plugin on public side.

dcWordTree 0.1.0 - 2014-12-01
==============================
* First public release


index.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014-2015 Nicolas Roudaire        http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

if (!defined('DC_CONTEXT_ADMIN')) { exit; }

$default_tab = 'settings';

$is_super_admin = $core->auth->isSuperAdmin();
$core->blog->settings->addNameSpace('dcwordtree');
$dcwordtree_active = $core->blog->settings->dcwordtree->active;
$dcwordtree_was_actived = $dcwordtree_active;

if (!empty($_POST['saveconfig'])) {
    try {
        $dcwordtree_active = (empty($_POST['dcwordtree_active']))?false:true;
        $core->blog->settings->dcwordtree->put('active', $dcwordtree_active, 'boolean');

        // change other settings only if they were in html page
        if ($dcwordtree_was_actived) {
            if (isset($_POST['dcwordtree_images'])) {
                $dcwordtree_images = explode("\n", trim($_POST['dcwordtree_images']));
                $core->blog->settings->dcwordtree->put('images', json_encode($dcwordtree_images), 'string');
            }
        }

        dcPage::addSuccessNotice(__('Configuration has been updated.'));
        http::redirect($p_url);
    } catch(Exception $e) {
        http::redirect($p_url);
    }
}

include(dirname(__FILE__).'/tpl/index.tpl');


_admin.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014 Nicolas Roudaire             http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

if (!defined('DC_CONTEXT_ADMIN')) { return; }

$_menu['Blog']->addItem(
    'dcWordTree',
    'plugin.php?p=dcWordTree',
    'index.php?pf=dcWordTree/imgs/dcWordTree.png',
    preg_match('/plugin.php\?p=dcWordTree/', $_SERVER['REQUEST_URI']),
    $core->auth->check('admin,contentadmin', $core->blog->id)
);


_public.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014-2015 Nicolas Roudaire        http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

if (!defined('DC_RC_PATH')) { return; }

$self_ns = $core->blog->settings->addNamespace('dcwordtree');
if (!$self_ns->active) {
    return;
}

$core->addBehavior('publicHeadContent', array('dcWordTreeBehaviors', 'publicHeadContent'));


_install.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014 Nicolas Roudaire             http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

$version = $core->plugins->moduleInfo('dcWordTree', 'version');
if (version_compare($core->getVersion('dcWordTree'), $version,'>=')) {
  return;
}

$settings = $core->blog->settings;
$settings->addNamespace('dcwordtree');

$settings->dcwordtree->put('active', false, 'boolean', 'dcWordTree plugin activated ?', false);

$core->setVersion('dcWordTree', $version);
return true;


_prepend.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014-2015 Nicolas Roudaire        http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

if (!defined('DC_RC_PATH')) { return; }

$__autoload['dcWordTreeBehaviors'] = dirname(__FILE__).'/inc/dcwordtree.behaviors.php';


js/dcwordtree.js
function draw(container, words, width) {
	var height = width*3/4;
	d3.select(container)
		.append("svg")
		.attr("width", width)
		.attr("height", height)
		.append("g")
		.attr("transform", "translate("+width/2+","+height/2+")")
		.selectAll("text")
		.data(words)
		.enter().append("text")
		.style("font-size", function(d) {
			return d.size + "px";
		})
		.style("font-family", "Impact")
		.style("fill", function(d, i) {
			return fill(i);
		})
		.attr("text-anchor", "middle")
		.attr("transform", function(d) {
			return "translate(" + [d.x, d.y] + ")rotate(" + d.rotate + ")";
		})
		.text(function(d) {
			return d.text;
		})
		.on("click", function(d) {
			window.location = d.url;
		});
}

$(function() {
	var words = [];
	$('#content .tags li a').each(function(i) {
		words.push({text: $(this).text(),
			    size: 20 + parseInt($(this).attr('class').replace(/^tag/, ''), 10) * 0.7,
			    url: $(this).attr('href')
			   });
	});

	$('#content .tags').hide();
	d3.layout.cloud().size([width, height])
		.words(words)
		.padding(padding)
		.rotate(function() { return ~~(Math.random() * 2) * 90; })
		.font("Impact")
		.fontSize(function(d) { return d.size; })
		.on("end", function() { draw('#content', words, $('#content').width());})
		.start();
});


inc/dcwordtree.behaviors.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014-2015 Nicolas Roudaire        http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

class dcWordTreeBehaviors
{
    public static function publicHeadContent($core) {
        $res = '';
        $res .= sprintf(
            '<script type="text/javascript" src="%s/js/d3.js"></script>'."\n",
            html::stripHostURL($core->blog->getQmarkURL().'pf=dcWordTree')
        );
        $res .= sprintf(
            '<script type="text/javascript" src="%s/js/d3.layout.cloud.js"></script>'."\n",
            html::stripHostURL($core->blog->getQmarkURL().'pf=dcWordTree')
        );
        $res .= sprintf(
            '<script type="text/javascript" src="%s/js/dcwordtree.js"></script>'."\n",
            html::stripHostURL($core->blog->getQmarkURL().'pf=dcWordTree')
        );
        $res .= '<script type="text/javascript">';
        $res .= 'var fill = d3.scale.category10();';
        $res .= 'var width = 800, height = 600, padding = 3;';
        $res .= '</script>';
        $res .= '<style type="text/css">';
        $res .= 'svg text { cursor: pointer; }';
        $res .= '</style>';

        echo $res;
    }
}

_define.php
<?php
// +-----------------------------------------------------------------------+
// | dcWordTree - a plugin for dotclear                                    |
// +-----------------------------------------------------------------------+
// | Copyright(C) 2014-2015 Nicolas Roudaire        http://www.nikrou.net  |
// +-----------------------------------------------------------------------+
// | This program is free software; you can redistribute it and/or modify  |
// | it under the terms of the GNU General Public License version 2 as     |
// | published by the Free Software Foundation                             |
// |                                                                       |
// | This program is distributed in the hope that it will be useful, but   |
// | WITHOUT ANY WARRANTY; without even the implied warranty of            |
// | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU      |
// | General Public License for more details.                              |
// |                                                                       |
// | You should have received a copy of the GNU General Public License     |
// | along with this program; if not, write to the Free Software           |
// | Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            |
// | MA 02110-1301 USA.                                                    |
// +-----------------------------------------------------------------------+

if (!defined('DC_RC_PATH')) { return; }

$this->registerModule(
	/* Name */			"dcWordTree",
	/* Description*/	"display tags in an amazing cloud",
	/* Author */		"Nicolas Roudaire",
	/* Version */		'0.2.0',
	/* Permissions */	"contentadmin",
	/* Properties */	array('type' => 'plugin',
							  'dc_min' => '2.6',
							  'support' => 'http://forum.dotclear.net/viewtopic.php?pid=333019',
							  'details' => 'http://plugins.dotaddict.org/dc2/details/dcWordTree'
    )
);
